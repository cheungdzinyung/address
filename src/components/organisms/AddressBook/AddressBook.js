import React, { useState } from "react";
import { ContactList } from "../ContactList";
import styles from "./AddressBook.module.css";

export const AddressBook = ({ contacts }) => {
  const [search, setSearch] = useState("");
  const filteredContacts =
    search.length === 0
      ? contacts
      : contacts.filter(
          ({ name }) =>
            name.last.toLowerCase().includes(search.toLowerCase()) ||
            name.first.toLowerCase().includes(search.toLowerCase())
        );

  return (
    <div className={styles.addressBook}>
      <input
        className={styles.searchBar}
        value={search}
        placeholder={"type to search by first or last name"}
        onChange={(e) => setSearch(e.target.value)}
      />
      <ContactList contacts={filteredContacts} />
    </div>
  );
};
