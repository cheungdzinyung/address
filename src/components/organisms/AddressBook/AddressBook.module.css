.addressBook {
  width: 80vw;
  margin: auto;
  padding: 30px 0;
}

.searchBar {
  width: 80%;
  background: white;
  padding: 10px;
}
