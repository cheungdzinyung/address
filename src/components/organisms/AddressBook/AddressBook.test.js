import React, { useState } from "react";
import { render, screen } from "@testing-library/react";
import { AddressBook } from "./AddressBook";
import Router from "react-router-dom";

const contacts = [
  {
    name: { first: "Andrew", last: "01" },
    picture: { large: "test 01" },
    phone: "phone111",
    cell: "cel111",
    login: { uuid: "1" },
  },
  {
    name: { first: "Andrew", last: "02" },
    picture: { large: "test 02" },
    phone: "phone222",
    cell: "cel222",
    login: { uuid: "2" },
  },
  {
    name: { first: "Test", last: "01" },
    picture: { large: "test 02" },
    phone: "phone222",
    cell: "cel222",
    login: { uuid: "3" },
  },
];

jest.mock("../ContactList", () => ({
  ContactList: ({ contacts }) => (
    <div>
      {contacts.map(({ login, name }) => (
        <div key={login.uuid}>{name.first}</div>
      ))}
    </div>
  ),
}));

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

const useStateMock = useState;

describe("AddressBook", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should render with 3 contacts", async () => {
    const mockSetState = jest.fn(() => "");

    useStateMock.mockImplementation((initialState) => [
      initialState,
      mockSetState,
    ]);

    const component = render(<AddressBook contacts={contacts} />);
    const inputElements = screen.getAllByRole("textbox");
    expect(inputElements.length).toBe(1);

    const contactElements = screen.queryAllByText(/Andrew/i);
    expect(contactElements.length).toBe(2);

    expect(component).toMatchSnapshot();
  });

  it("should render with 1 contact", async () => {
    //did not mock user input through input box, simply mocked state change
    const mockSetState = jest.fn();

    useStateMock.mockImplementation((initialState) => ["Test", mockSetState]);

    const component = render(<AddressBook contacts={contacts} />);
    const inputElements = screen.getAllByRole("textbox");
    expect(inputElements.length).toBe(1);

    const contactTextElements = screen.queryAllByText(/Test/i);
    expect(contactTextElements.length).toBe(1);

    const contactAndrewElements = screen.queryAllByText(/Andrew/i);
    expect(contactAndrewElements.length).toBe(0);

    expect(component).toMatchSnapshot();
  });
});
