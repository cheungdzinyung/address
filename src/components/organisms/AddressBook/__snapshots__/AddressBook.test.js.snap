// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressBook should render with 1 contact 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="addressBook"
      >
        <input
          class="searchBar"
          placeholder="type to search by first or last name"
          value="Test"
        />
        <div>
          <div>
            Test
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="addressBook"
    >
      <input
        class="searchBar"
        placeholder="type to search by first or last name"
        value="Test"
      />
      <div>
        <div>
          Test
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AddressBook should render with 3 contacts 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="addressBook"
      >
        <input
          class="searchBar"
          placeholder="type to search by first or last name"
          value=""
        />
        <div>
          <div>
            Andrew
          </div>
          <div>
            Andrew
          </div>
          <div>
            Test
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="addressBook"
    >
      <input
        class="searchBar"
        placeholder="type to search by first or last name"
        value=""
      />
      <div>
        <div>
          Andrew
        </div>
        <div>
          Andrew
        </div>
        <div>
          Test
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
